### PR TITLE
sharding: node shuffler implementation

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: sharding/xorValidatorShuffler_test.go [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+	for i := 0; i < number; i++ {
+		v[i] = &validator{
+			pubKey: generateRandomByteArray(32),
+		}
+	}
+
+	return v
+}
+
+func generateValidatorMap(
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: Suggestion when using random generated data: maybe should add a helper function that display a variadic parameter of  map[uint32][]Validator and use this function as the last parameter in asserts calls. The reason behind this request is: if, by chance, we find (or TC finds) an edge case, we can easily reproduce it because we have the values used. Otherwise, we need to run the same test over and over again hoping that the edge case is reproduced again.

File: sharding/xorValidatorShuffler_test.go [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+		result = append(result, valList...)
+	}
+
+	return result
+}
+
+func Test_xorBytes_SameLen(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(firstArray, secondArray)
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: maybe change rez to result, valid in the whole file.

File: sharding/xorValidatorShuffler_test.go [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+}
+
+func Test_copyValidatorMap(t *testing.T) {
+	t.Parallel()
+
+	valMap := generateValidatorMap(30, 2)
+	v2 := copyValidatorMap(valMap)
+	assert.Equal(t, valMap, v2)
+
+	valMap[0] = valMap[0][1:]
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: this can be replaced with `assert.False(t, valmap == v2)` as it tests pointer values

File: sharding/xorValidatorShuffler_test.go [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+func Test_shuffleList(t *testing.T) {
+	t.Parallel()
+
+	randomness := generateRandomByteArray(32)
+	validators := generateValidatorList(30)
+	validatorsCopy := make([]Validator, 0)
+	validatorsCopy = append(validatorsCopy, validators...)
+
+	shuffled := shuffleList(validators, randomness)
+	assert.Equal(t, len(validatorsCopy), len(shuffled))
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: :+1:

File: sharding/xorValidatorShuffler_test.go [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+	assert.True(t, contains(shuffled, validatorsCopy))
+}
+
+func Test_shuffleListParameterNotChanged(t *testing.T) {
+	t.Parallel()
+
+	randomness := generateRandomByteArray(32)
+	validators := generateValidatorList(30)
+	validatorsCopy := make([]Validator, 0)
+	validatorsCopy = append(validatorsCopy, validators...)
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: might use `copy(validatorsCopy, validators)`

File: sharding/xorValidatorShuffler_test.go [hunk 6]
```diff
@@ -0,0 +1,778 @@
...
+
+	_ = shuffleList(validators, randomness)
+	assert.Equal(t, validatorsCopy, validators)
+}
+
+func Test_shuffleListConsistentShuffling(t *testing.T) {
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: :+1:

File: sharding/xorValidatorShuffler_test.go [hunk 7]
```diff
@@ -0,0 +1,778 @@
...
+	validatorsCopy := make([]Validator, 0)
+	validatorsCopy = append(validatorsCopy, validators...)
+
+	shuffled := shuffleList(validators, randomness)
+	assert.Equal(t, len(validatorsCopy), len(shuffled))
+	assert.NotEqual(t, validatorsCopy, shuffled)
+	assert.True(t, contains(shuffled, validatorsCopy))
+}
+
+func Test_shuffleListParameterNotChanged(t *testing.T) {
```
Comment (iulianpascalau) [inline sharding/xorValidatorShuffler_test.go]: :+1:

File: sharding/xorValidatorShuffler_test.go (hunk 3 above)
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler_test.go]: I wanted to check deep equal not pointer value, copyValidatorMap does a deep copy.

File: sharding/xorValidatorShuffler.go [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+	rxs.mutShufflerParams.RUnlock()
+
+	if canSplit {
+		newEligible, newWaiting = rxs.splitShards(args.eligible, args.waiting, newNbShards)
+	}
+	if canMerge {
+		newEligible, newWaiting = rxs.mergeShards(args.eligible, args.waiting, newNbShards)
+	}
+
+	for shard, vList := range newWaiting {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler.go]: Actually the conditions are per shard, but maybe not very obvious. I will try to explain

If we would end up with less than the supported number of nodes for the current number of shards we would do the shard merge first and then proceed to remove the leaving from the resulting number of shards.

Let's say there are some nodes that announced their intent to leave.
Some of these nodes are in shard0, some in shard1, ..., some in metachain

From every shard we can remove the "leaving" nodes only if there are enough nodes left in the waiting list to take their place and not end up with a number of nodes below minimum accepted (nodesPerShard).

So the iteration is on each shard, search for each node in the leaving slice, and remove them if found as long as there is a node in the waiting list assigned to the shard.

Each node is unique so it can only be found in one shard.

There is a bias indeed for the first nodes in the leaving list, not in the first checked shard, but we can have the list shuffled with the same randomness if needed.

File: sharding/xorValidatorShuffler.go [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+	nbShardsNew := nbShards
+	if nodesNewEpoch > nodesForSplit {
+		nbNodesWithoutMaxMeta := nodesNewEpoch - maxNodesMeta
+		nbShardsNew = nbNodesWithoutMaxMeta / maxNodesShard
+
+		return nbShardsNew
+	}
+
+	if nodesNewEpoch == nodesForMerge {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler.go]: The merge will never happen in this case, as the nodes for merge will be the min number of nodes / shard multiplied by the number of shards and it can never go below.

I was thinking of adding another hysteresis threshold for merge that will allow some degree of variation.

File: sharding/xorValidatorShuffler_test.go (hunk 2 above)
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler_test.go]: done

File: sharding/xorValidatorShuffler_test.go (hunk 5 above)
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler_test.go]: done

File: sharding/xorValidatorShuffler.go [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+	newNbShards := rxs.computeNewShards(args.eligible, args.waiting, args.newNodes, args.leaving, args.nbShards)
+
+	rxs.mutShufflerParams.RLock()
+	canSplit := rxs.adaptivity && newNbShards > args.nbShards
+	canMerge := rxs.adaptivity && newNbShards < args.nbShards
+	rxs.mutShufflerParams.RUnlock()
+
+	if canSplit {
+		newEligible, newWaiting = rxs.splitShards(args.eligible, args.waiting, newNbShards)
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler.go]: the split and merge are still todo, but the way these will work will be the following:
Split/merge will set the new eligible and waiting lists based on the previous eligible and waiting lists:
- if merging is detected this round  eligible lists for the merging shards will be merged, waiting lists for the merging shards will be merged as well. 
- if split is detected and not prepared from prev epoch then prepare the split
- if split is detected and prepared then do the split by splitting equally the eligible and waiting lists.

Then nodes are shuffled out from eligible

Then the promotion of waiting list to eligible will be done (for a previously prepared split event the waiting lists should have enough nodes to fill up the split shards to required capacity)

the new nodes need to be equally distributed to the resulting shards new waiting lists

the shuffled out nodes are equally distributed to the resulting shards new waiting lists.

File: sharding/xorValidatorShuffler.go [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+	maxNodesMeta := rxs.nodesMeta + rxs.metaHysteresis
+	maxNodesShard := rxs.nodesShard + rxs.shardHysteresis
+	nodesForSplit := (nbShards+1)*maxNodesShard + maxNodesMeta
+	nodesForMerge := nbShards*rxs.nodesShard + rxs.nodesMeta
+	rxs.mutShufflerParams.RUnlock()
+
+	nbShardsNew := nbShards
+	if nodesNewEpoch > nodesForSplit {
+		nbNodesWithoutMaxMeta := nodesNewEpoch - maxNodesMeta
+		nbShardsNew = nbNodesWithoutMaxMeta / maxNodesShard
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler.go]: the extra nodes will be in waiting lists. the shards eligible lists will always be exactly of the desired capacity, never more or less.

File: sharding/xorValidatorShuffler.go [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+		nbToRemove := len(vList)
+		if len(args.leaving) < nbToRemove {
+			nbToRemove = len(args.leaving)
+		}
+
+		vList, args.leaving = removeValidatorsFromList(vList, args.leaving, nbToRemove)
+		newWaiting[shard] = vList
+	}
+
+	shuffledOutNodes, newEligible, args.leaving = shuffleOutNodes(args.eligible, args.waiting, args.leaving, args.rand)
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler.go]: right, the updated leaving list should have been returned.

File: sharding/xorValidatorShuffler.go [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+// The function needs to ensure that:
+//      1.  Old eligible nodes list will have up to shuffleOutThreshold percent nodes shuffled out from each shard
+//      2.  shuffledOutNodes = oldEligibleNodes + waitingListNodes - minNbNodesPerShard (for each shard)
+//      3.  Old waiting nodes list for each shard will be added to the remaining eligible nodes list
+//      4.  The leaving nodes are checked against the new nodes and shuffled out nodes and removed if present from the
+//          pools and leaving nodes list
+//      5.  The remaining new nodes are equally distributed among the existing shards into waiting lists
+//      6.  The remaining shuffled out nodes are distributed among the existing shards into waiting lists.
+//          We may have three situations:
+//          a)  In case (shuffled out nodes + new nodes) > (nbShards * perShardHysteresis + minNodesPerShard) then
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/xorValidatorShuffler.go]: right

File: sharding/interface.go [hunk 14]
```diff
@@ -46,3 +46,19 @@ type PublicKeysSelector interface {
 	GetValidatorsRewardsAddresses(randomness []byte, round uint64, shardId uint32) ([]string, error)
 	GetOwnPublicKey() []byte
 }
+
+// UpdateNodesArgs holds the parameters required by the shuffler to generate a new nodes configuration
+type UpdateNodesArgs struct {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/interface.go]: done

File: sharding/xorValidatorShuffler_test.go [hunk 15]
```diff
@@ -0,0 +1,778 @@
...
+	assert.Equal(t, validatorsCopy, validators)
+}
+
+func Test_shuffleListConsistentShuffling(t *testing.T) {
+	t.Parallel()
+
+	randomness := generateRandomByteArray(32)
+	validators := generateValidatorList(30)
+	validatorsCopy := make([]Validator, 0)
+	validatorsCopy = append(validatorsCopy, validators...)
```
Comment (golangcibot) [inline sharding/xorValidatorShuffler_test.go]: ineffectual assignment to `validatorsCopy` (from `ineffassign`)

File: sharding/xorValidatorShuffler_test.go (hunk 3 above)
Comment (iulianpascalau) [reply sharding/xorValidatorShuffler_test.go]: ok

File: sharding/xorValidatorShuffler.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/xorValidatorShuffler.go]: ok

File: sharding/xorValidatorShuffler.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/xorValidatorShuffler.go]: ok

File: sharding/xorValidatorShuffler.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/xorValidatorShuffler.go]: ok

File: sharding/xorValidatorShuffler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/xorValidatorShuffler.go]: ok

File: sharding/xorValidatorShuffler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/xorValidatorShuffler.go]: ok!

File: sharding/xorValidatorShuffler.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/xorValidatorShuffler.go]: ok
